bookdown/05-extending/05-extending-mlr3pipelines.Rmd

```diff
@@ -1,5 +1,470 @@
+    initialize = function(id = "copy.two") {
+      ....
+    },
+
+    train = function(inputs) {
+      ....
+    },
+
+    predict = function(inputs) {
+      ....
+    }
+  )
+)
+```
+
+#### Channel Definitions
+
+We need to tell the `PipeOp` the layout of its channels: How many there are, what their names are going to be, and what types are acceptable.
+This is done on initialization of the `PipeOp` (using a `super$initialize` call) by giving the `input` and `output` `data.table` objects.
+These must have three columns: a `"name"` column giving the names of input and output channels, and a `"train"` and `"predict"` column naming the class of objects we expect during training and prediction as input / output.
+A special value for these classes is `"*"`, which indicates that any class will be accepted; our simple copy operator accepts any kind of input, so this will be useful. We have only one input, but two output channels.
+
+By convention, we name a single channel `"input"` or `"output"`, and a group of channels [`"input1"`, `"input2"`, ...], unless there is a reason to give specific different names. Therefore, our `input` `data.table` will have a single row `<"input", "*", "*">`, and our `output` table will have two rows, `<"output1", "*", "*">` and `<"output2", "*", "*">`.
+
+All of this is given to the `PipeOp` creator. Our `initialize()` will thus look as follows:
+
+```{r 05-extending-017, eval = FALSE}
+    initialize = function(id = "copy.two") {
+      input = data.table::data.table(name = "input", train = "*", predict = "*")
+      # the following will create two rows and automatically fill the `train`
+      # and `predict` cols with "*"
+      output = data.table::data.table(
+        name = c("output1", "output2"),
+        train = "*", predict = "*"
+      )
+      super$initialize(id,
+        input = input,
+        output = output
+      )
+    }
+```
+
+#### Train and Predict
+
+Both `train()` and `predict()` will receive a `list` as input and must give a `list` in return.
+According to our `input` and `output` definitions, we will always get a list with a single element as input, and will need to return a list with two elements. Because all we want to do is create two copies, we will just create the copies using `c(inputs, inputs)`.
+
+Two things to consider:
+
+- The `train()` function must always modify the `self$state` variable to something that is not `NULL` or `NO_OP`. This is because the `$state` slot is used as a signal that `PipeOp`  has been trained on data, even if the state itself is not important to the `PipeOp` (as in our case). Therefore, our `train()` will set `self$state = list()`.
+
+- It is not necessary to "clone" our input or make deep copies, because we don't modify the data. However, if we were changing a reference-passed object, for example by changing data in a `Task`, we would have to make a deep copy first. This is because a `PipeOp` may never modify its input object by reference.
```

Comment:
Would be great if we could also stick to "one sentence one line" here. Just indent the sentence and the listing should be fine.